Download Helper added

khoin opened the pull request.

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,105 @@
+	}
+
+	// Prepare temporary directory
+	if err := os.MkdirAll("./temp/", os.ModePerm); err != nil {
```
Use ioutil.TempDir("","inmap")

https://golang.org/pkg/io/ioutil/#TempDir

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,105 @@
+
+	// Prepare temporary directory
+	if err := os.MkdirAll("./temp/", os.ModePerm); err != nil {
+		fmt.Fprintf(os.Stderr, "Failed creating temporary folder to download file. %v\n", err)
```
panic(fmt.Errorf("inmaputil: failed creating temporary download directory: %v", err)

ctessum replied:
We want to use a panic here because it's something that would happen because of a system error rather than a user error.

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,105 @@
+	"github.com/cavaliercoder/grab"
+)
+
+func Download(path string) string {
```
I don't think we want this function to be exported, so small d in download.

ctessum replied:
Include comments above the definition of all functions that follow best practices: https://blog.golang.org/godoc-documenting-go-code

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,105 @@
+	client := grab.NewClient()
+	req, _ := grab.NewRequest("./temp/", path)
+
+	fmt.Printf("Downloading %v...\n", req.URL())
```
Instead of writing log messages directly to the console, instead send them to a channel that's supplied by the caller, so the caller can decide what to do with them.

You can see an example implementation [here](https://github.com/spatialmodel/inmap/blob/master/run.go#L153), [here](https://github.com/spatialmodel/inmap/blob/master/run.go#L213), and [here](https://github.com/spatialmodel/inmap/blob/master/inmaputil/inmap.go#L136)

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,105 @@
+)
+
+func Download(path string) string {
+	// Check if local file exists
```
All comments within functions should be full sentences with capitals at the beginning and periods at the end.

ctessum commented on inmaputil/download_test.go:
```diff
@@ -0,0 +1,43 @@
+		t.Error("Expected /blah/test/, got ", k);
+	}
+
+	t.Logf("Passed directory tests. ");
```
This is probably not necessary because the testing framework will tell you it passes if it doesn't fail.

ctessum commented on inmaputil/download_test.go:
```diff
@@ -0,0 +1,43 @@
+	if test.Short() {
+		return
+	}
+	if k := Download("https://github.com/spatialmodel/inmap/blob/master/inmap/testdata/testEmis.shp.zip?raw=true"); k != "temp/testEmis.zip_extracted/testEmis.shp" {
```
We'll want to make sure this works in the case where there are multiple files in the zipfile and when the files are either within or not within a subdirectory in the shapefile.

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+	"github.com/cavaliercoder/grab"
+)
+
+// The download function checks if the input is an existing file locally.
```
download checks if...

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+// The download function checks if the input is an existing file locally.
+// If not, it checks if the file is a form of a URL.
+// If so, it downloads it and checks if the file needs to be decompressed.
+// It either returns the path of the Shp file or decompresses it and returns the path 
```
What's the difference between the two options in this sentence?

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+// If so, it downloads it and checks if the file needs to be decompressed.
+// It either returns the path of the Shp file or decompresses it and returns the path 
+// to the decompressed Shp file.
+// The second argument allows to specify a channel where logs will be written
```
c, if not nil, is a channel across which status update messages will be sent.

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+// to the decompressed Shp file.
+// The second argument allows to specify a channel where logs will be written
+func download(path string, c chan string) string {
+	// Check if local file exists. If it does, return the give path.
```
give -> given

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+		return path
+	}
+
+	// Preparing a temporary directory for the downloads.
```
Prepare a...

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	c <- fmt.Sprintf("Download saved to ./%v ", resp.Filename)
+	c <- ("Decompressing...")
```
We only want to say "decompressing" if we are actually decompressing a file. Currently it will always print.

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+
+	// choose the Shp file in the downloaded path.
```
Check if there is a file with the ".shp" extension in the extracted directory.
If so, return the path to it.

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+		if err != nil {
+			return resp.Filename
+		}
+	case ".gz":
```
This should be "tar.gz" instead of ".gz". ".gz" files without the ".tar" would just be a single compressed file.

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+	})
+
+	// We've walked the directory but found no Shp file
+	// So we return the downloaded path.
```
lowercase "s" in "so".

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+	// We've walked the directory but found no Shp file
+	// So we return the downloaded path.
+	if err != nil {
+		return resp.Filename
```
Also send a message over the channel saying what the error was.

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,122 @@
+		return foundShp;
+	}
+
+	close(c)
```
Probably delete this line; the caller may want to reuse the channel somewhere else.

khoin commented:
Hi Chris,

I've made a few changes from last pull request.

Also, I've tried to wrap some commands around `download()` in `inmaputil/cmd.go`, but I noticed that only "EmissionShapeFiles" matches the criteria to have `download()` wrapped around it. This is because the download() function looks for Shp files. 

Is this what you're looking for? Or does download.go has to be changed so that it complies not just with .shp files but other files?

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,123 @@
+		c <- ("Decompressing...")
+		err := archiver.Zip.Open(resp.Filename, resp.Filename + "_extracted")
+		if err != nil {
+			return resp.Filename
```
I think for this error and the similar ones below we should also send a message saying what the error is.

ctessum commented on inmaputil/download.go:
```diff
@@ -0,0 +1,123 @@
+
+	foundShp := "";
+
+	err = filepath.Walk(resp.Filename + "_extracted/", func(path string, info os.FileInfo,
```
I think we need to modify this function so that if it doesn't find a shapefile, it should return the path to the first file it finds.

ctessum commented:
We want to work for shapefiles and other files. I think it should currently work for non-shapefiles that are not compressed. I added a comment about how I think we can make it work files that are compressed but not shapefiles.

Also, there tests are failing because the download command also needs a channel as an argument: https://travis-ci.org/spatialmodel/inmap/jobs/355650570#L486

khoin commented:
I've made some changes. It passes 2/3 tests. Pass in Go 1.9.x but not 1.10.x. I will look into that. Still, there are some things you can look over and check if they're all right.

Particularly, I've added a helper function to create a `chan` that will output to whatever cobra.Command in the scope. I've also noticed that EmissionShapeFiles is an array and not a single file path, so I've wrapped `download` around each element. Let me know if that looks okay.

Additionally, here are some config variables which are defined, but isn't used anywhere, so I couldn't find and wrap `download` around them: 
VarGrid.CensusFile
VarGrid.MortalityRateFile

khoin commented:
I've changed the code a little bit and now it passes all the tests in Travis CI. It was simply a logging change and I have no clue why it's passing on Travis now.

However, it seems like some files in testdata were changed when I ran the tests and push the commit. I couldn't notice the files added/changed until I commit. I tried to download the repo ZIP and put the files back again then commit again but it still shows up in the Files Changed tab. Do you have any idea how this  could be mitigated?

Otherwise, just let me know if download.go, download_test.go, cmd.go is good and I'll just clone again and replace the files.

ctessum commented on inmaputil/cmd.go:
```diff
@@ -655,16 +666,23 @@ concentrations with no temporal variability.`,
 		}
+
+		shapeFiles := expandStringSlice(Cfg.GetStringSlice("EmissionsShapefiles"))
+		// wrap download over each filepaths
```
Full sentence.

ctessum commented on inmaputil/cmd.go:
```diff
@@ -655,16 +666,23 @@ concentrations with no temporal variability.`,
+		shapeFiles := expandStringSlice(Cfg.GetStringSlice("EmissionsShapefiles"))
+		// wrap download over each filepaths
+		for _, element := range shapeFiles {
+			element = download(element, createChan(cmd))
```
I'm not sure this would work as expected. You might need to do `shapeFiles[i] = download(...)`

ctessum commented on inmaputil/cmd.go:
```diff
@@ -655,16 +666,23 @@ concentrations with no temporal variability.`,
 			vgc,
-			os.ExpandEnv(Cfg.GetString("InMAPData")),
-			os.ExpandEnv(Cfg.GetString("VariableGridData")),
+			download(os.ExpandEnv(Cfg.GetString("InMAPData")), createChan(cmd)),
```
An issue with having the `createChan` function and then using it over and over again is that it ends up creating a bunch of channels and then never closing them. It might be better to create one channel and then use the same channel for all the download functions, and then close the channel when done with it.

ctessum commented:
I think git has `revert` and `reset` commands that can undo changes, but I don't remember the specifics. Google would probably know.

khoin commented:
I've made the changes, and cleaned my PR along with having some changes from `gopkg.lock` from `dep ensure`.

When I tested it locally and it passed, but Travis CI keeps failing for 1.10. I've also updated Golang on my local machine to 1.10 to test it out but it still passes locally here but not Travis. The Travis log isn't really helpful so I'll look into it later.

Can you see if it also works on your machine?

khoin commented:
I've removed the output files from this change so that they're not tracked.

Additionally, I've fixed the test files archive to include other files necessary for inmap to run.

I've also added a pause after the server is initiated ([download_test.go:34](https://github.com/spatialmodel/inmap/pull/33/files#diff-c8c47dacb1640abec5568d67b38d4fc0R34)) but Travis CI still fails the test (for the older Go version). At first, it showed up an error that the port is in use, so I've changed the port number to 12999 for testing now because it seems less likely now. (7777 seems like a nice number people people would use). It still fails for Go 1.9.x nonetheless (with the same invalid memory address error).

ctessum commented:
Hi Khoi,

A couple of comments:
* At least one of your errors is caused by passing a `nil` argument for the context in `srv.Shutdown`. It should instead be `srv.Shutdown(context.Background())`. After I change that, the tests run for me without panicking. It probably doesn't need a pause after the server initiation.
* You'll want to use the go auto-formatter to format your code before committing it. There are a number of editor plugins that do this automatically (e.g., [goclipse](http://goclipse.github.io/)), or you can do it with the [gofmt](https://golang.org/cmd/gofmt/) command.
* The port-in-use error is probably caused by running two instances of the tests at once, perhaps in two different terminals or one in an editor and one in the terminal.
* For `configExampleRemote.toml`, we probably want to have it try downloading all of the input files rather than just the emissions files. I think for the most part this will only require changes in the config file, but we will also need to create zip files for the population and mortality rate shapefiles.
* Change `testEmis.shp.zip` to `testEmis.zip`.
* In all cases, for documentation comments, use full sentences with Capitalized first letters and periods. (I understand that not all comments currently follow that rule, but we're trying to improve over time).

khoin commented:
Hi Chris,

* I've passed the context for the Shutdown method; it is without errors now. Thanks!
* I've installed gofmt as a plugin for Sublime Text. `gofmt` is now invoked every time I save the file. The changes are reflected in this commit.
* I haven't encountered the port-clashing error anymore so I haven't changed anything according to that.
* testEmis.shp.zip is renamed.

I've double checked the comments and have removed some that seems unnecessary. The existing ones are with the guidelines. Let me know if something needs to be changed. The exception I made to capitalizing the first letters is when the first word of the sentence is a variable/function name that is not capitalized. 

I will try and package the mortality and population rates into zip files and test it out later today. If there's a specific way you want me to package these files (e.g., separately, together, etc.), let me know.

khoin commented:
Hi Chris,

I've applied these changes, but I can not find anywhere in cmd.go to wrap `download()` around to load testPopulation.zip correctly. 

I suspect that although the field `VarGrid.CensusFile` is defined in `Cfg` variable in `cmd.go`, it is actually read through `VarGridConfig.loadPopulation()` [here](https://github.com/spatialmodel/inmap/blob/d55c6cf25d1d93cafe3c106344c4974d9b912549/vargrid.go#L796). 

A solution is to wrap `download()` around `config.CensusFile` in that line linked above, but that means I would also have to modify `vargrid.go`. Let me know if you approve of this.

The same situation is for MortalityRateFile as well.

ctessum commented:
Hi Khoi,

I think you can wrap the download function around the lines [here](https://github.com/spatialmodel/inmap/blob/master/inmaputil/config.go#L121) and [here](https://github.com/spatialmodel/inmap/blob/master/inmaputil/config.go#L124). Let me know if that doesn't work.

khoin commented:
Hi Chris,

The changes are reflected.

ctessum commented:
Fixes #32.

ctessum commented:
Thanks! One more comment: can you change the "download" function to "maybeDownload", so that it reflects the fact that it's only downloading files if they're not local?

Then, feel free to merge it.

ctessum commented:
I merged it. I don't know why this didn't close automatically.
